File: core/src/forward_index.cpp
#include "forward_index.hpp"

#include <iostream>

#include "isam_storage.hpp"
#include "lexicon.hpp"
#include "post.hpp"
#include "nlohmann/json.hpp"

void
ForwardIndex::generate(ISAMStorage& output_store,
                       ISAMStorage& data_index,
                       Lexicon& lexicon)
{
    std::vector<std::pair<uint64_t, std::string>> result;
    result.reserve(data_index.size());

    int c = 0;
    while (true) {
        auto p = data_index.next();
        if (!p.has_value()) break;

        std::string raw = p->second;

        nlohmann::json j = nlohmann::json::parse(raw);
        Post post = Post::from_json(j);

        std::string data = "";

        // Format:
        // wordID,Mask (space separated)
        //
        // Masks:
        // 1 = Title
        // 2 = Body
        // 3 = Tag

        // ---------- TITLE ----------
        if (post.post_type_id == 1) {
            auto t_tokens = Lexicon::tokenize_text(post.title);
            for (auto& t : t_tokens) {
                uint64_t wid = lexicon.get_word_id(t);
                data += std::to_string(wid) + ",1 ";
            }
        }

        // ---------- BODY ----------
        auto b_tokens = Lexicon::tokenize_text(post.cleaned_body);
        for (auto& t : b_tokens) {
            uint64_t wid = lexicon.get_word_id(t);
            data += std::to_string(wid) + ",2 ";
        }

        // ---------- TAGS ----------
        for (auto& t : post.tags) {
            uint64_t wid = lexicon.get_word_id(t);
            data += std::to_string(wid) + ",3 ";
        }

        result.emplace_back(p->first, data);

        std::cout << "Indexed " << (c + 1)
                  << " entries in forward index." << std::flush;

        c++;
    }
    std::cout << std::endl;

    std::cout << "Writing entries to disk";
    output_store.write(result);
    std::cout << "done." << std::endl;
}

File: Indexing/ForwardIndex.cs
using System;
using System.Collections.Generic;
using System.Text;

public static class ForwardIndex
{
    public static void Generate(ISAMStorage outputStore, ISAMStorage dataIndex, Lexicon lexicon)
    {
        // Use a smaller buffer to flush to disk periodically
        var batchBuffer = new List<(long Key, byte[] Data)>();
        const int BatchSize = 10000; 
        int count = 0;

        dataIndex.ResetNext();
        
        (long Key, byte[] Data)? entry;
        while ((entry = dataIndex.Next()) != null)
        {
            Post post = Post.Deserialize(entry.Value.Data);
            StringBuilder sb = new StringBuilder();

            // 1. Title (Mask 1)
            if (post.PostTypeId == 1 && !string.IsNullOrEmpty(post.Title))
            {
                foreach (var t in lexicon.Tokenize(post.Title))
                {
                    sb.Append($"{lexicon.GetWordId(t)},1 ");
                }
            }

            // 2. Body (Mask 2)
            if (!string.IsNullOrEmpty(post.CleanedBody))
            {
                foreach (var t in lexicon.Tokenize(post.CleanedBody))
                {
                    sb.Append($"{lexicon.GetWordId(t)},2 ");
                }
            }

            // 3. Tags (Mask 4)
            foreach (var t in post.Tags)
            {
                sb.Append($"{lexicon.GetWordId(t)},4 ");
            }

            // Convert to bytes and add to buffer
            byte[] encodedData = Encoding.UTF8.GetBytes(sb.ToString().TrimEnd());
            batchBuffer.Add((entry.Value.Key, encodedData));

            count++;

            // Periodically flush to the ISAM file to save RAM
            if (count % BatchSize == 0)
            {
                outputStore.InsertBatch(batchBuffer);
                batchBuffer.Clear();
                Console.Write($"\rIndexed {count} entries...");
            }
        }

        // Final flush for remaining items
        if (batchBuffer.Count > 0)
        {
            outputStore.InsertBatch(batchBuffer);
        }

        Console.WriteLine($"\nIndexing Complete. Total: {count}");
    }
}